### ISSUE TITLE
Ambiguous "unsafe" Error WASM JS alert

### ISSUE THREAD

**Issue by dgb23** [rust-analyzer/rust-analyzer#5412]
I'm in the process of following the Game of Life Rust-WASM tutorial here:

https://rustwasm.github.io/docs/book/game-of-life/hello-world.html

My level of expertise in Rust: Beginner

OS: macOS Catalina 10.15.5
Editor: VSCode 1.47
rust-analyzer: v0.2.240

(rust-analyzer updated just a few minutes ago)

---

I'm getting this error message when looking at src/lib.rs:

```
{
	"resource": "[...]/wasm-game-of-life/src/lib.rs",
	"owner": "_generated_diagnostic_collection_name_#1",
	"severity": 8,
	"message": "This operation is unsafe and requires an unsafe function or block",
	"source": "rust-analyzer",
	"startLineNumber": 18,
	"startColumn": 5,
	"endLineNumber": 18,
	"endColumn": 16
}
```

In this code block the *second* `alert(...)` is highlighted:

```
#[wasm_bindgen]
extern {
    fn alert(s: &str);
}

#[wasm_bindgen]
pub fn greet(name: &str) {
    alert(name);
}
```

**However the code compiles regardless.**

When changing to:

```
#[wasm_bindgen]
pub fn greet(name: &str) {
    unsafe{alert(name);}
}
```

Then I'm getting this warning:

```
{
	"resource": "[...]/wasm-game-of-life/src/lib.rs",
	"owner": "_generated_diagnostic_collection_name_#1",
	"code": "unused_unsafe",
	"severity": 4,
	"message": "unnecessary `unsafe` block\n`#[warn(unused_unsafe)]` on by default",
	"source": "rustc",
	"startLineNumber": 18,
	"startColumn": 5,
	"endLineNumber": 18,
	"endColumn": 11
}
```

**Comment by jonas-schievink** [comment]
That's just because we don't expand procedural macros yet. Presumably it does some magic to make these operations safe.

**Comment by dgb23** [comment]
> That's just because we don't expand procedural macros yet. Presumably it does some magic to make these operations safe.

Would you rather recommend to get the "unsafe" error or the "unused unsafe" warning? Or in other words is my temporary 'fix' the right thing to do or should I just ignore it?

**Comment by jonas-schievink** [comment]
I'd suggest listening to the rustc warning, not rust-analyzer. If the wrong diagnostics from rust-analyzer annoys you, you can also turn it off in the settings (but this currently turns *all* of them off).

**Comment by kinnison** [comment]
I just hit this today myself.  Is there a timeline on rust-analyzer being able to run proc-macros?

**Comment by bjorn3** [comment]
Derive proc macros already work when you enable them. The other kinds are not yet supported. I don't know of any timeline for imolementing this.

**Comment by jonas-schievink** [comment]
The latest release has preliminary support for function-like procedural macros like `foo!();`.

Attribute macros are harder to implement and require pretty hairy name resolution changes.

**Comment by FermiDirak** [comment]
For those looking to disable the `missing-unsafe` rule in VS Code until it's fixed, adding the following to your `settings.json` and reloading VS Code will suppress the error:

```
"rust-analyzer.diagnostics.disabled": [
    "missing-unsafe"
]
```

**Comment by digama0** [comment]
Is it possible to disable these diagnostics locally, using an attribute like with clippy lints?

**Comment by jonas-schievink** [comment]
Not yet, no

**Comment by OliverEvans96** [comment]
I just ran into this issue with [cxx](https://github.com/dtolnay/cxx), which uses macros to generate safe C++ FFIs.

**Comment by bjorn3** [comment]
You can enable proc-macro support using

```jsonc
{
    "rust-analyzer.cargo.loadOutDirsFromCheck": true,
    "rust-analyzer.procMacro.enable": true,
}
```

**Comment by OliverEvans96** [comment]
Thanks @bjorn3, but unfortunately, that didn't get rid of the erroneous warning. I think I'll stick with disabling `missing-unsafe` for now.

**Comment by lnicola** [comment]
@OliverEvans96 `cxx` also uses attribute proc macros, which are not supported.

**Comment by FrankenApps** [comment]
You can get rid of this error / warning locally by putting an unsafe block and then surpressing the warning, but it is not a great way to handle that:
```rust
#[wasm_bindgen]
pub fn greet(name: &str) {
    #[allow(unused_unsafe)]
    unsafe {
        alert(&format!("Hello, {}!", name));
    }
}
```

**Comment by PanopticaRising** [comment]
Just in case anyone doesn't think to check the merged PR and runs into this, I can confirm that I was able to set this in VS Code and got the warning to go away:

```
    "rust-analyzer.experimental.procAttrMacros": true,
```

**Comment by fosskers** [comment]
For folks using Doom Emacs, this seems to do the trick:
```lisp
(after! lsp-rust
  (setq lsp-rust-analyzer-proc-macro-enable t
        lsp-rust-analyzer-experimental-proc-attr-macros t))
```

**Comment by lnicola** [comment]
Recent `rust-analyzer` versions should have both of those options enabled.
